web/app/fr/gospeak/web/pages/orga/partners/contacts/ContactCtrl.scala

```diff
@@ -0,0 +1,132 @@
+package fr.gospeak.web.pages.orga.partners.contacts
+
+import java.time.Instant
+
+import cats.data.OptionT
+import cats.effect.IO
+import com.mohiva.play.silhouette.api.Silhouette
+import com.mohiva.play.silhouette.api.actions.SecuredRequest
+import fr.gospeak.core.domain.{Contact, Group, Partner}
+import fr.gospeak.core.services.storage.{ContactRepo, OrgaGroupRepo, OrgaPartnerRepo, OrgaUserRepo}
+import fr.gospeak.libs.scalautils.Extensions._
+import fr.gospeak.libs.scalautils.domain.Page
+import fr.gospeak.web.auth.domain.CookieEnv
+import fr.gospeak.web.auth.exceptions.DuplicateEmailException
+import fr.gospeak.web.domain.Breadcrumb
+import fr.gospeak.web.pages.orga.contacts._
+import fr.gospeak.web.pages.orga.partners.PartnerCtrl
+import fr.gospeak.web.pages.orga.partners.contacts.ContactCtrl._
+import fr.gospeak.web.pages.orga.partners.contacts.routes.{ContactCtrl => ContactRoutes}
+import fr.gospeak.web.pages.orga.partners.routes.{PartnerCtrl => PartnerRoutes}
+import fr.gospeak.web.utils.Mappings._
+import fr.gospeak.web.utils.UICtrl
+import play.api.data.Form
+import play.api.data.Forms.{mapping, optional}
+import play.api.mvc.{Action, AnyContent, ControllerComponents, Result}
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+class ContactCtrl(cc: ControllerComponents,
+                  silhouette: Silhouette[CookieEnv],
+                  contactRepo: ContactRepo,
+                  userRepo: OrgaUserRepo,
+                  groupRepo: OrgaGroupRepo,
+                  partnerRepo: OrgaPartnerRepo
+                 ) extends UICtrl(cc, silhouette) {
+
+  import silhouette._
+
+  private val createForm: Form[Contact.Data] = Form(mapping(
+    "partner" -> partnerId,
+    "first_name" -> firstName,
+    "last_name" -> lastName,
+    "email" -> emailAddress,
+    "description" -> optional(description)
+  )(Contact.Data.apply)(Contact.Data.unapply))
+
+
+  def list(group: Group.Slug, partner: Partner.Slug, params: Page.Params): Action[AnyContent] = SecuredAction.async { implicit req =>
+    (for {
+      groupElt <- OptionT(groupRepo.find(user, group))
+      partnerElt <- OptionT(partnerRepo.find(groupElt.id, partner))
+      contacts <- OptionT.liftF(contactRepo.list(partner, params))
+      b = listBreadcrumb(groupElt, partnerElt)
+    } yield Ok(html.list(groupElt, partnerElt, contacts)(b))).value.map(_.getOrElse(groupNotFound(group))).unsafeToFuture()
```

Comment:
```suggestion
    } yield Ok(html.list(groupElt, partnerElt, contacts)(b))).value.map(_.getOrElse(partnerNotFound(group, partner))).unsafeToFuture()
```
Is more appropriated, the function will redirect to the group page and if group is not found, the group page will redirect to the user page

Reply:
Done